docs/proposals/scheduling/20230314-gpu-topology-aware-scheduling.md

```diff
@@ -0,0 +1,275 @@
+  - health: true
+    id: GPU-3680858f-1753-371e-3c1a-
+    minor: 1
+    resources:
+      kubernetes.io/gpu-core: "100"
+      kubernetes.io/gpu-memory: 16Gi
+      kubernetes.io/gpu-memory-ratio: "100"
+    type: gpu
+  - health: true
+    id: GPU-95fe7a8b-bf9b-73cc-2903-c6e65883f3a7
+    minor: 2
+    resources:
+      kubernetes.io/gpu-core: "100"
+      kubernetes.io/gpu-memory: 16Gi
+      kubernetes.io/gpu-memory-ratio: "100"
+    type: gpu
+  - health: true
+    id: GPU-cd8d5d8c-7334-4c68-587e-04202daa38a5
+    minor: 3
+    resources:
+      kubernetes.io/gpu-core: "100"
+      kubernetes.io/gpu-memory: 16Gi
+      kubernetes.io/gpu-memory-ratio: "100"
+    type: gpu
+  - health: true
+    id: GPU-511dd579-5044-b716-e08a-841f51796a59
+    minor: 4
+    resources:
+      kubernetes.io/gpu-core: "100"
+      kubernetes.io/gpu-memory: 16Gi
+      kubernetes.io/gpu-memory-ratio: "100"
+    type: gpu
+  - health: true
+    id: GPU-62460a09-6838-abc8-00f5-31d2c6c101ef
+    minor: 5
+    resources:
+      kubernetes.io/gpu-core: "100"
+      kubernetes.io/gpu-memory: 16Gi
+      kubernetes.io/gpu-memory-ratio: "100"
+    type: gpu
+  - health: true
+    id: GPU-2da27328-f395-a226-7486-c08e6e98570f
+    minor: 6
+    resources:
+      kubernetes.io/gpu-core: "100"
+      kubernetes.io/gpu-memory: 16Gi
+      kubernetes.io/gpu-memory-ratio: "100"
+    type: gpu
+  - health: true
+    id: GPU-8137b226-b69a-1f22-4367-da110c8ba6b5
+    minor: 7
+    resources:
+      kubernetes.io/gpu-core: "100"
+      kubernetes.io/gpu-memory: 16Gi
+      kubernetes.io/gpu-memory-ratio: "100"
+    type: gpu
+```
+#### Node Selection
+Suppose a training job has a total of 3 pods waiting to be scheduled, namely pod1, pod2, and pod3. When pod1 is scheduled, the list of available nodes filtered by the filter extension point is [node1, node2, node3]. The logic for selecting available node groups for [pod1, pod2, pod3] is as follows:
+
+- First try to use one node to place [pod1, pod2, pod3]. The conditions for a node to place these three pods are as follows:
+   - The number of GPUs available on the candidate node is greater than or equal to the sum of the number of GPUs requested by [pod1, pod2, pod3].
+   - Call the RunFilterPlugins function provided by the scheduler on the candidate node and each pod to run all filter extension points to determine whether the pod can run on the node. If all pods can run on the node, then the candidate node can place the set of pods:
+```
+// nodeInfo is the current node info
+satisfied := true 
+for _,pod := range []*v1.Pod{pod1,pod2,pod3} {
+	status := RunFilterPlugins(context.TODO(), state, pod, nodeInfo)
+	if status.Merge().IsSuccess() {
+			nodeInfo.AddPod(p)
+	}else {
+    satisfied := false
+  	break
+  }
+}
+```
+
+- If one node cannot place [pod1, pod2, pod2], then try to place these three pods with 2 nodes. After allocating GPUs to the pods, the combination with less remaining GPU resources on the node is preferred
```

Comment:
if there 10 pod of a job, and node only can fill 2 pod, so will you try every arrange and combine? like pod1+pod2 or pod1 + pod3 or ....? (C2/100).  may be we can assume that all pod in a gpu job that need nvlink is Isomorphic(actually it's indeed Isomorphic in real world), this will lead the problem much easier.  we want this semanteme of  "A node can place 2 pod, B node can place 1 pod...",  not "A node can place pod1+pod3, node2 can place pod2"...
"

Reply:
if pod1\pod2\pod3 and node1\node2\node3,  the arrange and combine is [pod1\pod2 on node1, pod1\pod3 on node1...], [pod1\pod2 on node2, pod1\pod3 on node2...], which determined by pod num and node num, which is terrible,  the process may we first calculate each node's max can-assign num, and sort the node, then place the Isomorphic pod to the node from max->min